### PR TITLE
Bugfix

### PR DESCRIPTION
**Pull request by smarshall-spitzbart**
# Description

Bugfix

## Linked issues

Intentionally not closing any issues with this PR. Applicable issues will be closed in base branch.

- [x] `Fix`: Changes and/or adds code behavior, specifically to fix a bug
- [x] `Testing`: Adds testing

## New behavior tests

New test that validates fix: `TestPacketSpam`

### CODE REVIEW

File: x/ccv/provider/keeper/throttle.go [hunk 1]
```diff
@@ -19,54 +19,56 @@ import (
...
-		if !found || val.IsJailed() {
-			// If validator is not found, or found but jailed, it's power is 0. This path is explicitly defined since the
-			// staking keeper's LastValidatorPower values are not updated till the staking keeper's endblocker.
-			valPower = 0
-		} else {
-			valPower = k.stakingKeeper.GetLastValidatorPower(ctx, val.GetOperator())
-		}
+	// Obtain all global slash entries, where only some of them may be handled in this method,
+	// depending on the value of the slash meter.
+	allEntries := k.GetAllGlobalSlashEntries(ctx)
```
Comment (mpoke) [inline x/ccv/provider/keeper/throttle.go:29]: My concern here is that the size of this slice depends on the number of slash packets received from consumer chains. This is related to https://github.com/cosmos/interchain-security/issues/594. A solution here would be to have a filter on the getter that only returns the entries the throttling mechanism can handled.

File: x/ccv/provider/keeper/throttle.go [hunk 2]
```diff
@@ -19,54 +19,56 @@ import (
...
-		// Store handled global entry to be deleted after iteration is completed
-		handledGlobalEntries = append(handledGlobalEntries, entry)
-
-		// Do not handle anymore global slash entries if the meter is negative in value
-		stop = meter.IsNegative()
-		return stop
-	})
+		// don't handle any more global entries if meter becomes negative in value
+		if meter.IsNegative() {
+			break
```
Comment (mpoke) [inline x/ccv/provider/keeper/throttle.go:43]: Include this break into `GetAllGlobalSlashEntries` as meter is not modified in `HandlePacketDataForChain`. This also means that `handledEntries` is no longer needed.

File: x/ccv/provider/keeper/throttle.go [hunk 3]
```diff
@@ -19,54 +19,56 @@ import (
...
 	// Handled global entries are deleted after iteration is completed
-	k.DeleteGlobalSlashEntries(ctx, handledGlobalEntries...)
+	k.DeleteGlobalSlashEntries(ctx, handledEntries...)
 
 	// Persist current value for slash meter
 	k.SetSlashMeter(ctx, meter)
 }
 
+// Obtains the effective validator power relevant to a validator consensus address.
+func (k Keeper) GetEffectiveValPower(ctx sdktypes.Context,
```
Comment (mpoke) [inline x/ccv/provider/keeper/throttle.go:55]: Safe to use in an iterator as it only contains getters.

**Review by mpoke — COMMENTED** [review]
Amazing work @smarshall-spitzbart. I left some comments that can be dealt with in another PR after we merge https://github.com/cosmos/interchain-security/pull/462.

**Comment by smarshall-spitzbart** [comment]
@mpoke I think your comments here can be addressed by an explanation. 

If we were to add a filter into `GetAllGlobalSlashEntries` which called `GetEffectiveValPower`, this would essentially have to be the old design with two iterators open. Reasoning:

The value of `GetEffectiveValPower` is different before and after a validator is jailed. Think about this example. 

> Slash meter starts at value 5, 3 slash packets are recv and queued by provider, each requesting to jail the same validator with a voting power of 4. 

With your proposed design, we would obtain 2 slash packets from store where the 3rd would be filtered out (5-4=+1, 1-4 = -3). This means we would not handle the 3rd slash packet, when we indeed could have. 

With current design (and also the double-iterator design) we would obtain and handle all 3 slash packets, since the first slash packet would be handled, the validator jailed, then the subsequent calls to `GetEffectiveValPower` would return 0.

Lastly, `GetEffectiveValPower` contains getters to an external keeper, which do not contain iterators right now, but there's a small chance they could be refactored in the future to contain an iterator. Therefore the function is safe to use in an iterator right now, but could silently change in the underlying code it calls. 

This is an example of where more effective iterator usage would be desirable for performance reasons, but would be dangerous and considered bad practice. It seems like iterator filtering functions should pretty much always be stateless, including reading from state

**Comment by smarshall-spitzbart** [comment]
TLDR: if we don't get all global slash entries, it'll make #594 worse

**Comment by mpoke** [comment]
> With your proposed design, we would obtain 2 slash packets from store where the 3rd would be filtered out (5-4=+1, 1-4 = -3). This means we would not handle the 3rd slash packet, when we indeed could have.

@smarshall-spitzbart That's true. Didn't think of that. 

Then, we really need to make sure that we don't add invalid packets to that queue, i.e., #594
